pkg/provisioner/nacos/provisioner.go

```diff
@@ -0,0 +1,319 @@
+	"github.com/api7/apisix-mesh-agent/pkg/provisioner/util"
+	"github.com/api7/apisix-mesh-agent/pkg/types"
+	"github.com/api7/apisix-mesh-agent/pkg/types/apisix"
+)
+
+var (
+	_ provisioner.Provisioner = (*nacosProvisioner)(nil)
+
+	// ErrUnknownNacosScheme means provided nacos address is not HTTP or HTTPS
+	ErrUnknownNacosScheme = errors.New("invalid nacos scheme")
+)
+
+// TODO: Add e2e test
+type nacosProvisioner struct {
+	clientConfig  *constant.ClientConfig
+	serverConfigs []constant.ServerConfig
+
+	logger *log.Logger
+
+	// nacos config client
+	configClient config_client.IConfigClient
+
+	// eventsCh receive changes from client
+	eventsCh chan []types.Event
+	// mu should be acquire before update routes or upstreams
+	mu sync.Mutex
+	// last state of routes
+	routes []*apisix.Route
+	// last state of upstreams
+	upstreams []*apisix.Upstream
+}
+
+func NewProvisioner(cfg *config.Config) (provisioner.Provisioner, error) {
+	clientLogLevel := cfg.LogLevel
+	switch clientLogLevel {
+	case "dpanic", "panic", "fatal": // valid in zap but invalid in nacos-go
+		clientLogLevel = "error"
+	default:
+		clientLogLevel = "info"
+	}
+
+	clientConfig := constant.NewClientConfig(
+		constant.WithNotLoadCacheAtStart(true),
+		constant.WithLogDir("/tmp/nacos/log"),
+		constant.WithCacheDir("/tmp/nacos/cache"),
+		constant.WithRotateTime("1h"), // Log rotate time
+		constant.WithLogLevel(clientLogLevel),
+	)
+
+	if !strings.Contains(cfg.NacosSource, "://") {
+		cfg.NacosSource = "http://" + cfg.NacosSource
+	}
+	u, err := url.Parse(cfg.NacosSource)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return nil, ErrUnknownNacosScheme
+	}
+
+	var port int
+	if u.Port() != "" {
+		port, err = strconv.Atoi(u.Port())
+		if err != nil {
+			return nil, err
+		}
+	} else if u.Scheme == "http" {
+		port = 80
+	} else if u.Scheme == "https" {
+		port = 443
+	}
+
+	serverConfigs := []constant.ServerConfig{
+		{
+			IpAddr:      u.Hostname(),
+			ContextPath: u.Path,
+			Port:        uint64(port),
+			Scheme:      u.Scheme,
+		},
+	}
+
+	logger, err := log.NewLogger(
+		log.WithOutputFile(cfg.LogOutput),
+		log.WithLogLevel(cfg.LogLevel),
+		log.WithContext("nacos-provisioner"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &nacosProvisioner{
+		clientConfig:  clientConfig,
+		serverConfigs: serverConfigs,
+		eventsCh:      make(chan []types.Event),
+		logger:        logger,
+	}, nil
+}
+
+func (p *nacosProvisioner) Channel() <-chan []types.Event {
+	return p.eventsCh
+}
+
+func (p *nacosProvisioner) Run(stop chan struct{}) error {
+	defer close(p.eventsCh)
+
+	configClient, err := clients.NewConfigClient(
+		vo.NacosClientParam{
+			ClientConfig:  p.clientConfig,
+			ServerConfigs: p.serverConfigs,
+		},
+	)
+	if err != nil {
+		return err
+	}
+	p.configClient = configClient
+
+	err = p.fullSync()
+	if err != nil {
+		return err
+	}
+
+	err = p.watch()
+	if err != nil {
+		return err
+	}
+
+	for {
```

Comment:
Don't need this for loop.

Reply:
A missing `unwatch` func is added.

Reply:
But this is not related to the outer `for` loop.